stablehlo/reference/NumPy.cpp

```diff
@@ -0,0 +1,344 @@
+                                     "Failed to open NumPy file.");
+
+    llvm::raw_fd_ostream out(fd, /*shouldClose=*/true);
+
+    buildNumpyHeader<T>(out, type.getShape());
+    out.write(data, sizeof(T) * type.getNumElements());
+
+    return llvm::Error::success();
+  }
+};
+
+template <typename T>
+class FromNumpy {
+ public:
+  llvm::ErrorOr<Tensor> operator()(StringRef filename, ShapedType type) {
+    std::ifstream in(filename.str(), std::ifstream::binary);
+    size_t wordSize = 0;
+
+    if (readNumpyHeader<T>(in, wordSize, type.getShape()))
+      return llvm::errc::invalid_argument;
+
+    const int64_t numLoadedElements = std::accumulate(
+        type.getShape().begin(), type.getShape().end(), 1, std::multiplies<>());
+
+    if (numLoadedElements != type.getNumElements() || wordSize != sizeof(T))
+      return llvm::errc::invalid_argument;
+
+    const size_t dataBytesToRead = wordSize * numLoadedElements;
+    std::vector<T> data(dataBytesToRead);
+    in.read(reinterpret_cast<char*>(data.data()), dataBytesToRead);
+
+    return Tensor(type,
+                  HeapAsmResourceBlob::allocateAndCopyInferAlign<T>(data));
+  }
+};
+}  // namespace
+
+template <template <typename Type> class Functor, typename... Args>
+static decltype(auto) dispatchType(Type type, Args&&... args) {
+  if (auto quantized = type.dyn_cast<quant::QuantizedType>()) {
+    switch (quantized.getStorageTypeIntegralWidth()) {
+      case 8:
+        return Functor<int8_t>()(std::forward<Args>(args)...);
+      case 16:
+        return Functor<int16_t>()(std::forward<Args>(args)...);
+      default:
+        llvm::report_fatal_error("Unhandled quantized type");
+    }
+  }
+  if (type.isSignlessInteger(1))
+    return Functor<char>()(std::forward<Args>(args)...);
+  if (type.isSignlessInteger(8))
+    return Functor<int8_t>()(std::forward<Args>(args)...);
+  if (type.isInteger(8)) return Functor<uint8_t>()(std::forward<Args>(args)...);
+  if (type.isSignlessInteger(16))
+    return Functor<int16_t>()(std::forward<Args>(args)...);
+  if (type.isInteger(16))
+    return Functor<uint16_t>()(std::forward<Args>(args)...);
+  if (type.isSignlessInteger(32))
+    return Functor<int32_t>()(std::forward<Args>(args)...);
+  if (type.isInteger(32))
+    return Functor<uint32_t>()(std::forward<Args>(args)...);
+  if (type.isSignlessInteger(64))
+    return Functor<uint64_t>()(std::forward<Args>(args)...);
+  if (type.isInteger(64))
+    return Functor<int64_t>()(std::forward<Args>(args)...);
+  if (type.isF16()) return Functor<uint16_t>()(std::forward<Args>(args)...);
+  if (type.isF32()) return Functor<float>()(std::forward<Args>(args)...);
+  if (type.isF64()) return Functor<double>()(std::forward<Args>(args)...);
+  if (auto complexTy = type.dyn_cast<ComplexType>()) {
+    auto complexElemTy = complexTy.getElementType();
+    if (complexElemTy.isF32())
+      return Functor<double>()(std::forward<Args>(args)...);
+
+#ifdef __SIZEOF_INT128__
```

Comment:
For proper `f64` complex type serialization, we'd need a 128 bit native type to properly go to/from disk (64 bits for each component). Since it's not standardized, I think we need a preprocessor check here. We can however use a standard `double` for `f32` complex types. Added a comment.